proposals/0055-csp.md

```diff
@@ -0,0 +1,461 @@
+>
+```
+
+
+#### `Astro.csp.insertScriptHash(string)`
+
+```js
+Astro.csp.insertScriptHash("sha256-qewpiorfuvng3")
+```
+Result
+```html
+<meta 
+  http-equiv="Content-Security-Policy"
+  content="style-src 'self' 'sha256-hash1'; script-src 'self' 'sha256-hash2' 'sha256-qewpiorfuvng3'"
+>
+```
+
+#### `Astro.csp.setScriptResource(string)`
+
+Astro will deduplicate and remove duplicated resources.
+
+```js
+Astro.csp.setScriptResource("https://scripts.cdn.example.com")
+```
+Result
+```html
+<meta 
+  http-equiv="Content-Security-Policy"
+  content="style-src 'self' 'sha256-hash1' 'sha256-qewpiorfuvng3'; script-src https://scripts.cdn.example.com 'sha256-hash2'"
+>
+```
+
+
+# Testing Strategy
+
+Types of tests:
+- unit tests
+- integration tests
+- e2e tests
+
+Features to test:
+- client islands
+- server islands
+- view transitions
+- external scripts (via APIs)
+- external styles (via APIs)
+- inline scripts
+- inline styles
+
+# Drawbacks
+
+The hashes solution will negatively impact the rendering engine because Astro will have to calculate the hashes **at runtime**. The most expensive solution is the support of server islands because the contents of their scripts depend on props and slots, which are only known during the rendering phase.
+
+However, users are willing to compromise some performance penalties for security.
+
+## Dev server
+
+`vite` works differently in dev and prod. In dev, all assets (scripts and styles) are inlined, and also replaced using HRM. After the build, these assets are bundled, minified and imported differently.
+
+Due to this significant discrepancy, supporting the dev server will be very difficult, so we will focus on the build first.
+
+# Alternatives
+
+## The `nonce` solution
+
+This solution is way **easier** to implement; however, it won't work for static pages (SSG).
+As highlighted in this [comment of mine](https://github.com/withastro/roadmap/issues/1149#issuecomment-2824552131), in order to provide the `nonce` header, a user would need to use **an edge function** in a serverless environment or middleware for server environments. This means that environments that don't provide any virtual host control (e.g. GitHub pages, etc.), the solution wouldn't work *at all*.
+
+This solution doesn't follow Astro core principles, where an Astro site should work *out of the box* with the majority of environments, especially SSG, which is a first-class citizen for us.
+
+
+# Adoption strategy
+
+- The usage of this feature will be under an experimental flag, and shipped in a minor:
+
+  ```js
+  export default defineConfig({
+      experimental: {
+          csp: true
+      }
+  })
+  ```
+
+- The new Astro feature will be shipped with prefix `_exprimental`: `_experimentalStaticHeaders` in a patch release.
+- The new field of the hook `astro:build:generated` will be shipped with prefix `_experimental`: `_experimentalRouteToHeaders` in a patch release.
```

Comment:
When you make a PR against `next` to stabilize CSP, we gotta remember to update these names

Reply:
`_experimentalRouteToHeaders` was added for CSP, but it's been designed to be generic. Which means that the two features aren't directly connected, and they can become stable independently. If by the time of stabilisation, `_experimentalRouteToHeaders` is still named like this, we shouldn't change it.